source/RoslynAnalyzers/MessageContractAnalyzers.cs

```diff
@@ -0,0 +1,222 @@
+                        }
+                    }
+                }
+
+                if (requestOrCommandDec != null || responseEventOrResourceDec != null)
+                {
+                    // this is an "API Surface" type; either (request, command, response) or event or resource
+                    // note technically everything else in the Octopus.Server.MessageContracts namespace is also an "API surface" type, 
+                    // but verifying that would be more expensive and we don't need to do it yet
+                    ApiContractTypes_MustLiveInTheAppropriateNamespace(context, typeDec);
+                    
+                    if (requestOrCommandDec != null || responseEventOrResourceDec?.Identifier.Text == TypeNameIResponse)
+                    {
+                        // this is a "MessageType"; either request, command, or response
+                        CheckProperties(context, typeDec);
+                        MessageTypes_MustHaveXmlDocComments(context, typeDec);
+                    }
+                }
+                
+                // request/command/response specific
+                switch (requestOrCommandDec?.Identifier.Text)
+                {
+                    case TypeNameICommand:
+                        if (CommandTypes_MustBeNamedCorrectly(context, typeDec))
+                            CommandTypes_MustHaveCorrectlyNamedResponseTypes(context, typeDec, requestOrCommandDec); // the expected response name depends on the command name; so only run this check if the CommandName was good
+                        break;
+
+                    case TypeNameIRequest:
+                        if (RequestTypes_MustBeNamedCorrectly(context, typeDec))
+                            RequestTypes_MustHaveCorrectlyNamedResponseTypes(context, typeDec, requestOrCommandDec); // the expected response name depends on the request name; so only run this check if the RequestName was good
+                        break;
+                }
+
+                if (responseEventOrResourceDec?.Identifier.Text == TypeNameIEvent)
+                {
+                    EventTypes_MustBeNamedCorrectly(context, typeDec);
+                }
+            }
+        }
+        
+        static bool CheckProperties(SyntaxNodeAnalysisContext context, TypeDeclarationSyntax typeDec)
+        {
+            var result = true;
+            foreach (var propDec in typeDec.DescendantNodes().OfType<PropertyDeclarationSyntax>())
+            {
+                // only validate public properties
+                if (!propDec.Modifiers.Any(m => m.IsKind(SyntaxKind.PublicKeyword))) continue;
+
+                // we have a number of places where we treat collection types differently; front-load that info
+                var isCollectionType = IsCollectionType(context, propDec);
+                var required = GetRequiredState(propDec);
+
+                // if anything returns false, propagate false outwards
```

Comment:
Not entirely sure what you mean here.
I suspect you mean "only run the next on if the previous one succeeded"

to my mind, the `result &=`, approach isn't that common. I'd prefer a less terse but easier to read `if (!check(...)) return;` approach.

Reply:
>I suspect you mean "only run the next on if the previous one succeeded"

Not quite; The job this is doing is "run everything and report if any errors were found; no short-circuiting".
The binary & is equivalent to a sequence of

```
if(!check1(...)) result = false;
if(!check2(...)) result = false;
...
```

Reply:
Ah - makes sense. I think that adds weight to the lets go for a less terse syntax approach :)
But... Not a hill i'm going to die on.